vendor/blockchain_contracts/src/ethereum/rfc003/erc20_htlc.rs

```diff
@@ -0,0 +1,219 @@
+        "6000000000000000000000000000000000000006";
+
+    const DEPLOY_HEADER_TEMPLATE: &'static str =
+        include_str!("./templates/out/erc20_deploy_header.asm.hex");
+    const CONTRACT_START_POSITION_PLACEHOLDER: &'static str = "1001";
+    const CONTRACT_LENGTH_PLACEHOLDER: &'static str = "2002";
+
+    pub fn new(
+        refund_timestamp: Timestamp,
+        refund_address: Address,
+        redeem_address: Address,
+        secret_hash: SecretHash,
+        token_contract_address: Address,
+        amount: U256,
+    ) -> Self {
+        Self {
+            refund_timestamp,
+            refund_address,
+            redeem_address,
+            secret_hash,
+            token_contract_address,
+            amount,
+        }
+    }
+
+    /// Constructs the payload for funding an `Erc20` HTLC located at the given
+    /// address.
+    pub fn funding_tx_payload(&self, htlc_contract_address: Address) -> Bytes {
+        let transfer_fn_abi = base16!("A9059CBB");
+        let htlc_contract_address = <[u8; 20]>::from(htlc_contract_address);
+        let amount = <[u8; 32]>::from(self.amount);
+
+        let mut data = [0u8; 4 + 32 + 32];
+        data[..4].copy_from_slice(transfer_fn_abi);
+        data[16..36].copy_from_slice(&htlc_contract_address);
+        data[36..68].copy_from_slice(&amount);
+
+        Bytes::from(data.to_vec())
+    }
+
+    pub fn token_contract_address(&self) -> Address {
+        self.token_contract_address
+    }
+
+    pub fn deployment_gas_limit(&self) -> U256 {
+        let bytes: Bytes = self.compile_to_hex().into();
+        let n_bytes = bytes.0.len();
+        let gas_per_byte = 200;
+
+        U256::from(100_000 + n_bytes * gas_per_byte)
+    }
+
+    pub fn tx_gas_limit() -> U256 {
+        U256::from(100_000)
+    }
+
+    pub fn fund_tx_gas_limit() -> U256 {
+        U256::from(100_000)
+    }
+
+    pub fn compile_template_to_hex() -> String {
+        let code_length = Self::CONTRACT_CODE_TEMPLATE.len() / 2; // In hex, each byte is two chars
+
+        let code_length_as_hex = format!("{:0>4x}", code_length);
+
+        let header_length = Self::DEPLOY_HEADER_TEMPLATE.len() / 2;
+        let header_length_as_hex = format!("{:0>4x}", header_length);
+
+        let deploy_header = Self::DEPLOY_HEADER_TEMPLATE
+            .to_string()
+            .replace(
+                Self::CONTRACT_START_POSITION_PLACEHOLDER,
+                &header_length_as_hex,
+            )
+            .replace(Self::CONTRACT_LENGTH_PLACEHOLDER, &code_length_as_hex);
+
+        deploy_header + Self::CONTRACT_CODE_TEMPLATE
+    }
+
+    fn get_offset(name: &str, placeholder: &str) -> Offset {
+        let contract = Self::compile_template_to_hex();
+
+        let re_match = Regex::new(placeholder)
+            .expect("Could not create regex")
+            .find(contract.as_str())
+            .expect("Could not find placeholder in hex code");
+        Offset::new(
+            name.into(),
```

Comment:
:wrench: 

This `name` is a bit dodgy, especially because it contains formatting information (tabs). Maybe you could make use of an enum to express the different placeholders + the information that is associated with them?

Reply:
I knew you were going to say that but haven't thought of something better. I'll look into the Enum.

Reply:
(see my comment below. I don't think we need an enum).

Reply:
done

Reply:
The idea with the enum would have been to express _all_ the information associated with those placeholders:
- placeholder name
- a function for replacing the placeholder
- the name of the placeholder etc

At the moment, there is a lot of duplication and expanded "iterations" in regards to the placeholders and their use. I voiced the idea with the enum because it seemed you preferred actual iteration over the current expanded one (`Vec<Offset>`). I'd prefer if we are not mixing styles here:

Either use Iterators to go through all the placeholders and perform the logic or always flatten them out. I'd actually prefer the 2nd option because
1. the rest of the code is done like this
2. the list is not expected to grow and explicitly referencing the placeholders makes it more visible, what we are doing here